### ISSUE TITLE
Runtime bug: Hostage integer underflow

### ISSUE THREAD

**Issue by 1da1a172** [dxx-rebirth/dxx-rebirth#648]
### Environment
Arch Linux, updated on 2022-07-28.
Installed from AUR; d2x-rebirth-git-0.60.0.beta2.r1861.gd1ed53ebd-1

I _think_ this corresponds to commit d1ed53ebd9cb9bad753c5e25331a184461338d9e.

#### CPU environment
* [ ] x86 (32-bit Intel/AMD)
* [x] x86\_64 (64-bit Intel/AMD)
* [ ] ARM (32-bit)
* [ ] ARM64 (64-bit; sometimes called AArch64)

#### Game environment

Observable in Descent 2: Counterstrike, Level 14 and Secret Level 4.

It could be recreated in a custom campaign where there are hostages in a secret level which can be exited without blowing the reactor.

### Description

If you rescue all the hostages in the main level and at least 1 hostage from a secret level, without blowing up the reactor in the secret level, then you exit the main level with more hostages on board than there were in the mine. This leads to an integer underflow. This leads to about 4.2 billion hostages lost on the scoring screen, and no full rescue bonus.

#### Regression status

I am unaware of what versions this is present in. I wouldn't be surprised if it were present in the original game.

### Steps to Reproduce

Note, this is a bug in the scoring, so you will need to do this without cheats. I don't have a level editor, but it should be pretty straight forward to create a minimum test case.

You will need a level, and a secret level with hostages, that you can exit with the hostages, without blowing the reactor in the secret level. In the main campaign, Level 14 and Secret Level 4 allow for this. In the secret level, you will need a guide missile to open the locked door from the other side. See https://youtube.com/clip/UgkxdAVLJwdlKMxJpwsjNc40jkyANQ66N_t3 for an example.

Rescue rescue hostages from the base level and the secret level, but don't blow the reactor in the secret level. You don't need to rescue all the hostages, but you do need to get a total that is more than there are in the base level (L14 has 4 hostages, SL4 has 5). Complete the base level, and successfully exit the mine.

#### Expected behavior

If all the hostages from the base level and the secret level are rescued, then the full rescue bonus should be awarded, and no hostages should be lost.

I'm not sure how the hostages from the secret level should be counted. Presumably, any that make it out after blowing the reactor in the main level are rescued. Any that are left behind are not _dead_ but if there are no portals to the secret level in future levels, then they are _unreachable_.

#### Observed behavior

Playing on Hostshot, a full rescue of L14 and SL4 resulted in a hostage bonus of 13500, 4294967291 hostages lost, and no full rescue bonus. Note that the number of hostages lost is 0xfffffffb, or -5 in an unsigned 32 bit int. This corresponds to the number of extra hostages we left the mine with.

Pretty clearly, what is happening is that the hostage count is reset when a mine is destroyed (or a player dies). The scoring at the end of the game does:
(total number of hostages in the mine) - (hostages onboard) = number of hostages left in the mine
If that number is 0, then the full rescue bonus is awarded. If not, then it is the number of hostages lost. We've just managed to legitimately rescue more hostages than there were in the mine.

This problem does not occur in D1, as the mechanics of the game do not allow for leaving a mine (even a secret level) without destroying the reactor.

* Engines affected:
  * [ ] D1X-Rebirth
  * [x] D2X-Rebirth

* Game modes affected:
  * [x] Single player
  * [ ] Multiplayer competitive
    * [ ] Anarchy
    * [ ] Team Anarchy
    * [ ] Robo-Anarchy
    * [ ] Capture the Flag
    * [ ] Bounty
  * [ ] Multiplayer cooperative

**Comment by vLKp** [comment]
This was previously indirectly reported in pull request #497, though that report failed to call out that there are stock levels which trigger this problem.  The reporter of that issue proposed simply ignoring the underflow.  I pushed back, asserting that the secret level hostages should be properly accounted.  The reporter never responded.  The request is still open.

The way secret levels are handled is messy, so I am not surprised there are weird cases like this.

**Comment by 1da1a172** [comment]
All of the solutions that I can think of have issues, but pretty much anything is better than an integer underflow.

As a start, maybe we could alter the logic. Instead of trying to find out how many hostages were lost by counting things at the end, include a field for how many hostages were lost due to the player dying, and check for any hostages still in the mine.

 It still leaves questions open for what is even the correct way to score hostages in secret levels. What is the intent on scoring secret levels, especially in regards to hostages? Are bonuses awarded at all? I don't think there is a score screen when destroying a secret level. Should there be? Does that actually help the situation?

I don't suppose there is a mechanism for this project to modify the stock levels? I can think of some minor tweaks that would avoid the issue (without significantly altering the game play), and leave custom level designers to ensure the same. That doesn't seem likely, so probably a moot point.

**Comment by 1da1a172** [comment]
I just did some play testing, and even when blowing the reactor on the secret level, I still got an integer underflow. Now I have no idea how to keep this from happening.

**Comment by AlumiuN** [comment]
Yeah, that would be expected with the current behaviour - it always takes the expected total from the base level only, even if you have more hostages.

For what it's worth, I think the best solution would be to ignore hostages from secret levels. Trying to determine at any given point as to whether the bonus should apply when you leave a level is going to be an absolute nightmare otherwise (as even in the base game, despite the original intention of the developers, you can leave D2 secret level 4 with all of the hostages without blowing the reactor).

**Comment by vLKp** [comment]
The problem with adding any additional accounting is that there is no reliable way to save and restore that information, due to limitations of the savegame format.  I could add fields to it, but older builds will blindly assume those fields don't exist, and may behave strangely as a result.  Thus, any new accounting needs to be done through clever use of the existing fields, such as what I suggested in that old pull request: adjust the "count of hostages on level" field up on exiting a secret level where hostages were saved.

I don't think adding a score screen on leaving the secret level is quite right either, since you're not necessarily forced to destroy the level on exit.

Looking back at how older releases of Rebirth handled this, the underflow did not happen, but the player would miss the "Full rescue" bonus because the rescued count was not equal to the count of hostages on the base level.

**Comment by 1da1a172** [comment]
What was the behavior from the original game? That may provide some guidance on expected behavior.

Not having secret level hostages increase the number of hostages on board actually seems like a reasonable way to cheese it.

Loading a save that is newer than the build of the game seems unlikely. The other way around is sure to happen on occasion just with upgrades, but this way requires sharing a save to someone running an older version of the game. I'd be OK with that risk. You can always add a version check, as well. That doesn't help current builds, but it gives some safety to modify other stuff in the future.

**Comment by vLKp** [comment]
As I read the code, the behavior of the original game is as I described above:

https://github.com/dxx-rebirth/dxx-rebirth/blob/f8bc08f08331ce51b296f4ddec78f292ce12bb95/main/gameseq.c#L1140

That link leads to shortly after the Descent 2 source was imported into a public source control system.  It is not exactly what was used for the retail build, but is likely close enough for this purpose, particularly since it still agrees with the much later behavior seen in 0.58.x.

---

As regards ignoring secret level hostages, does that mean that the player is forbidden to rescue the hostages on secret levels, or just that the player gets a "full rescue" bonus if the player rescued _at least as many_ hostages as the base level had, regardless of whether that was achieved by ignoring the secret level hostages and clearing the main level, ignoring the main level hostages and clearing the secret level, or some combination thereof?  I have no objection to using a `>=` test instead of strict `==` as the original game did.  I believe when I made the change that caused this problem, I thought that the count of rescued hostages could never exceed the count of total hostages, and so the change was a simple optimization, rather than a behavior change.
